sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java

```diff
@@ -128,27 +128,10 @@ public static String escapeStringLiteral(final String s)
   }
 
   /**
-   * Convert {@link RelDataType} to the most appropriate {@link ValueType}, coercing all ARRAY types to STRING (until
-   * the time is right and we are more comfortable handling Druid ARRAY types in all parts of the engine).
-   *
-   * Callers who are not scared of ARRAY types should isntead call {@link #getValueTypeForRelDataTypeFull(RelDataType)},
-   * which returns the most accurate conversion of {@link RelDataType} to {@link ValueType}.
+   * Convert {@link RelDataType} to the most appropriate {@link ValueType}}.
    */
   @Nullable
   public static ColumnType getColumnTypeForRelDataType(final RelDataType type)
```

Comment:
I take it that no callers are "scared of ARRAY types" anymore.

1. What are the main risks of making this change, & are they mitigated by testing?
2. Do we need a feature flag?

Reply:
the engine has been waiting on array grouping for this coercion to go away, though there is a missing piece I think where we need to ensure that when grouping on a virtual column we never plan to a topn query in `DruidQuery`.

I think most of the `MV_` prefixed functions have their return type as `VARCHAR` rather than `ARRAY`, so in theory the "documented" multi-value string functions shouldn't actually be affected, and any that will be is a bug that we should fix. The undocumented `ARRAY_` prefixed functions will now behave as intended and validate/operate as true SQL arrays (and so group correctly as arrays). If anyone was using the undocumented `ARRAY_` prefix functions, then I guess that behavior will change, since grouping on those expressions would previously have been grouping as `STRING` due to the implicit unnest.

The only reason I can think of to add a feature flag is if need a way to hide bugs that this change causes, and that _should_ only happen if any of the `MV_` functions have bugs in their SQL output types, or if people were using the undocumented `ARRAY_` functions.

Reply:
As part of this patch, then, let's check all the MV functions to make sure the return types are what we expect them to be.

As to the array stuff, one of them is documented: ARRAY_AGG. And you could do that in a subquery and group on the result. Some questions about that:

- Do we have tests for the case of grouping on the result of an ARRAY_AGG? If not we should add it.
- Does the behavior change or does this introduce excess risk due to potential bugs? If so, a feature flag would be a good idea.

Reply:
> As part of this patch, then, let's check all the MV functions to make sure the return types are what we expect them to be.

Looks like all mv functions are overriding the output signature of the corresponding array functions to return a VARCHAR instead of an array

> 
> As to the array stuff, one of them is documented: ARRAY_AGG. And you could do that in a subquery and group on the result. Some questions about that:
> 
>     * Do we have tests for the case of grouping on the result of an ARRAY_AGG? If not we should add it.
Were you looking for [this](https://github.com/apache/druid/pull/12078/files#diff-8bc53aec44924e671b3c6f6f34c6f0c499f873d9000649c47c237444707aea4bL1640) test case ?

> 
>     * Does the behavior change or does this introduce excess risk due to potential bugs? If so, a feature flag would be a good idea.

Still thinking about this

Reply:
Adding a flag in the query Context and take a decision based on that may not be trivial as the code will need to be extensively refactored.
For eg : checkout `RowSignature#fromRelDataType` https://github.com/apache/druid/blob/master/sql/src/main/java/org/apache/druid/sql/calcite/table/RowSignatures.java
I am not sure if adding changes to a few places where the QueryContext is passable and where its not passable, convert to array without honoring the flag, has on the query Engine. Especially in the case where the user has set the flag and half of the places we are honoring the flag and the other half we are not.

Therefore adding a RuntimeProperty in `ExpressionProcessingConfig`

Reply:
> Were you looking for [this](https://github.com/apache/druid/pull/12078/files#diff-8bc53aec44924e671b3c6f6f34c6f0c499f873d9000649c47c237444707aea4bL1640) test case ?

Yes! thanks!